### ISSUE TITLE
patch: declare that twisted requires pywin32 if it is to offer process management or iocp reactor on Windows

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3238]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)| @zooko reported|
|-|-|
|Trac ID|trac#3238|
|Type|enhancement|
|Created|2008-05-15 22:39:11Z|

See also distutils-sig discussion:

http://mail.python.org/pipermail/distutils-sig/2008-May/009501.html

```
Index: setup.py
===================================================================
--- setup.py    (revision 23632)
+++ setup.py    (working copy)
@@ -75,6 +75,8 @@
     if 'setuptools' in sys.modules:
         from pkg_resources import parse_requirements
         requirements = ["zope.interface"]
+        if sys.platform == 'win32':
+            requirements.append('pywin32')
         try:
             list(parse_requirements(requirements))
         except:
```

Attachments:

* [twistedrequirespywin32.patch.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6ca/6ca1e9a46879877b8c4309ee9c4bc773fbb6e468/7ca55027c90b3121e91940dfd831425fcdbbed12.txt) (458 bytes) - added by zooko on 2008-05-15 22:39:31Z - patch against svn trunk to setup.py
* [svndiff.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6ca/6ca1e9a46879877b8c4309ee9c4bc773fbb6e468/feb7ef7ab1f82ab6305aa303b2a9d0084eb0655d.txt) (987 bytes) - added by zooko on 2008-06-10 01:27:00Z - twisted requires pywin32 if on win32 to offer iocp-reactor or process-management
* [svndiff.2.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6ca/6ca1e9a46879877b8c4309ee9c4bc773fbb6e468/13891cd91912dc1381a4c70ddb9063e0cb34f30f.txt) (987 bytes) - added by zooko on 2008-11-13 22:07:48Z - if using setuptools and installing on Windows, declare that Twisted needs pywin32 in order to provide process management or iocp reactor
* [twisted-master.cfg](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6ca/6ca1e9a46879877b8c4309ee9c4bc773fbb6e468/a1b2e91a7712067c84fac1a8d38fa598f992bd03.cfg) (26939 bytes) - added by exarkun on 2009-09-20 00:19:01Z - Twisted's master.cfg
* [divmod-master.cfg](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6ca/6ca1e9a46879877b8c4309ee9c4bc773fbb6e468/2fa53ededeba8e2e30f5975f4541a948f171bfd3.cfg) (30215 bytes) - added by exarkun on 2009-09-20 00:21:34Z - Divmod's master.cfg
* [twisted_factories.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/6ca/6ca1e9a46879877b8c4309ee9c4bc773fbb6e468/ea4c1b3eec9764c4e22bba13cf78ec54275d65e9.py) (29214 bytes) - added by exarkun on 2011-06-09 11:20:32Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3238 3238
type__enhancement enhancement
reporter__zooko zooko
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__release_management release management
keywords__ 
time__1210891151000000 1210891151000000
changetime__1503984128221856 1503984128221856
version__None None
owner__ 
cc__therve cc__zooko
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

fixing description markup

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve set owner to @zooko|
|-|-|

Actually, pywin32 is not required, or should not be. I think it's required by the iocp reactor, and process management, but you should be able to run a twisted application without that.

What make you think it's required?

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_3"></a>|@zooko commented|
|-|-|

Hm.  Thinking about it, I guess I must have thought that it was required because the various projects that I know of that use Twisted on Windows all say that pywin32 is required.  Perhaps that is because they all use the iocp reactor by default?  I believe that the non-iocp reactor has limitations on the number of simultaneous open sockets which make it unusable for many projects.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

It's probably more linked with process management than reactor. The default select reactor handles 1024 sockets, so it's enough for lots of application. And IOCP is still a movable target.

So I'd say it's more a requirement to put on applications that uses Twisted AND iocp/process than in Twisted itself. -1 for me.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_5"></a>|@zooko commented|
|-|-|

It does?  I thought the default reactor failed after 32 fds.

I agree that this is an application requirement -- that twisted itself on Windows does not require pywin32.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_6"></a>|@zooko commented|
|-|-|

We could use the "extras" keyword to declare that Twisted requires pywin32 if it is running on Windows and the user wants to use process management or iocp reactor, something like this:

```
Index: setup.py
===================================================================
--- setup.py    (revision 23635)
+++ setup.py    (working copy)
@@ -75,6 +75,10 @@
     if 'setuptools' in sys.modules:
         from pkg_resources import parse_requirements
         requirements = ["zope.interface"]
+        extras_requirements = {}
+        if sys.platform == 'win32':
+            extras_requirements['process_management'] = ['pywin32']
+            extras_requirements['iocp_reactor'] = ['pywin32']
         try:
             list(parse_requirements(requirements))
         except:
@@ -84,6 +88,7 @@
 """
         else:
             setup_args['install_requires'] = requirements
+            setup_args['extras_require'] = extras_requirements
         setup_args['include_package_data'] = True
         setup_args['zip_safe'] = False
     setup(**setup_args)
```

Now someone who does "easy_install Twisted", or whose requirements list "Twisted", will not get Twisted installed, but if their requirements list a dependency on "Twisted [iocp_reactor]", then they *will* get pywin32 if they install on Windows.  Here is the documentation about this feature:

http://peak.telecommunity.com/DevCenter/setuptools#declaring-extras-optional-features-with-their-own-dependencies

I've tested this patch on my machine and it didn't cause any problem.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_7"></a>|@zooko commented|
|-|-|

I wrote "Now someone who does "easy_install Twisted", or whose requirements list "Twisted", will not get Twisted installed" but I meant that they would not get pywin32 installed.  Of course they would get Twisted installed.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_8"></a>|@zooko commented|
|-|-|

Okay, it turns out that a feature like this will be necessary if Tahoe is going to be able to automatically install Twisted on Windows and get iocp reactor and process management.  If Twisted would please apply a patch like the one in the message above, then Tahoe could express the fact that it requires "Twisted with the option of using iocp reactor" or "Twisted with process management", by putting something like the following in the Tahoe metadata:

```
install_requires['Twisted[iocp_reactor,process_management]']
```

Then users of Tahoe who install Tahoe on Windows will, in addition to automatically getting Twisted installed so that Tahoe can use it, also automatically get pywin32 installed so that Twisted can use that.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve commented|
|-|-|

Note that easy-installing twisted on windows now needs a C compiler to work. Maybe we could provide some eggs, but we don't for now, so I'm not sure setting the dependency on pywin32 is a high priority. Why not simply declare it in you setup.py?

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_10"></a>|@zooko commented|
|-|-|

I don't understand the part about requiring a C compiler.  What requires a C compiler?

"Why not simply declare it in you setup.py?"

That would work, but I think it is better for Twisted to declare it because it is actually a requirement of Twisted (when you want iocp reactor or process management behavior from Twisted) rather than of Tahoe.  For example, perhaps a future version of Twisted will be released that provides iocp reactor without pywin32.  Then instead of all packages which use Twisted with iocp reactor changing their setup.py's to say something like "If Twisted >= 8.6 then we don't require pywin32 else we do.", Twisted can show that it supplies the iocp reactor to its users while no longer requiring pywin32.  Basically the whole issue of whether or not you require pywin32 should be up to Twisted to configure properly instead of up to every package which uses Twisted.

Thank you for your time.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve commented|
|-|-|

Replying to [zooko](#note_10):
> I don't understand the part about requiring a C compiler.  What requires a C compiler?

iocp needs a pyrex extension.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_12"></a>|@zooko commented|
|-|-|

I see.  So, having Twisted declare that, if the user wants iocp reactor, then Twisted will require pyrex at build time and pywin32 at installation, does not make it harder for people to install Twisted.

It does not make it so that they require a C compiler when previously they didn't.  Rather it formalizes the information of whether or not a package which requires Twisted also requires that stuff.  Make sense?

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_13"></a>|@zooko commented|
|-|-|

Here's the port of this patch to the current trunk -- [25344].  (It is the exact same patch but shifted by a couple of lines.)

Please apply.

```
HACM yukyuk:~/playground/twisted/trunk$ svn diff
Index: setup.py
===================================================================
--- setup.py    (revision 25344)
+++ setup.py    (working copy)
@@ -78,6 +78,13 @@
     if 'setuptools' in sys.modules:
         from pkg_resources import parse_requirements
         requirements = ["zope.interface"]
+        extras_requirements = {}
+        if sys.platform == 'win32':
+            extras_requirements['process management'] = ['pywin32']
+            extras_requirements['iocp reactor'] = ['pywin32']
+        else:
+            extras_requirements['process management'] = []
+            extras_requirements['iocp reactor'] = []
         try:
             list(parse_requirements(requirements))
         except:
@@ -87,6 +94,7 @@
 """
         else:
             setup_args['install_requires'] = requirements
+            setup_args['extras_require'] = extras_requirements
         setup_args['include_package_data'] = True
         setup_args['zip_safe'] = False
     setup(**setup_args)
```

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_14"></a>|@zooko commented|
|-|-|

clarify Summary:

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

Any thoughts on testing strategy for this change?

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_16"></a>|@zooko commented|
|-|-|

Well, we want to check that:

(a) if your project requires iocp_reactor or process_management functionality from Twisted on Windows that your projects setup realizes that this means it requires pywin32.

(b) the patch doesn't break anything else.

Here is a minimal setup.py:

```
from setuptools import setup

setup(name="minimal_project_to_test_twisted_extra",
    install_requires=["Twisted[iocp_reactor"]
)
```

If you execute this with ```mkdir -p instdir/lib/python2.5/site-packages ; PYTHONPATH=instdir/lib/python2.5/site-packages python ./setup.py install --prefix=instdir```, then it will tell you that it can't install because Twisted doesn't offer the "iocp_reactor" functionality.  If you apply the patch to Twisted to declare that it does offer the iocp_reactor functionality then invoking this ```setup.py``` script should succeed.

How's that for a start?

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_17"></a>|@zooko commented|
|-|-|

Hi, I was just testing out the process of installing Tahoe on Windows.  It currently fails unit tests with this error message:

```
[Failure instance: Traceback: <type 'exceptions.NotImplementedError'>: spawnProcess not available since pywin32 is not installed.
```

If you accept the patch in this ticket, then I can move that issue from run-time to install-time, express it in a programmable and consistent way with all of the other Tahoe dependencies (see http://allmydata.org/trac/tahoe/browser/_auto_deps.py ), and even have the dependency on pywin32 automatically resolved by our install process.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_18"></a>|@zooko commented|
|-|-|

I just tried this patch against current trunk and it still works for me.  I guess it is a bad idea to use spaces in names of packagey things like extras, so this patch changes the names from "process management" and "iocp reactor" to "process_management" and "iocp_reactor".

```
Index: setup.py
===================================================================
--- setup.py    (revision 26987)
+++ setup.py    (working copy)
@@ -78,6 +79,13 @@
     if 'setuptools' in sys.modules:
         from pkg_resources import parse_requirements
         requirements = ["zope.interface"]
+        extras_requirements = {}
+        if sys.platform == 'win32':
+            extras_requirements['process_management'] = ['pywin32']
+            extras_requirements['iocp_reactor'] = ['pywin32']
+        else:
+            extras_requirements['process_management'] = []
+            extras_requirements['iocp_reactor'] = []
         try:
             list(parse_requirements(requirements))
         except:
@@ -87,8 +95,18 @@
 """
         else:
             setup_args['install_requires'] = requirements
+            setup_args['extras_require'] = extras_requirements
         setup_args['include_package_data'] = True
         setup_args['zip_safe'] = False
```

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_19"></a>|@zooko commented|
|-|-|

I noticed this ticket: [#3397](https://github.com/twisted/twisted/issues/3397) which ends with this note from Glyph:

"""
Since Twisted is functioning properly here I'm closing this ticket. However, perhaps you can suggest a way to make this less confusing for other users and open a new ticket for better documentation. 
"""

I propose this ticket ([#3238](https://github.com/twisted/twisted/issues/3238)) as a way to make this less confusing for other users and even to serve as a kind of documentation for Python-literate users.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_20"></a>|@zooko commented|
|-|-|

So I think that this ticket is blocked on someone implementing a test for it.  comment:17 is my suggestion for how such a test would work.  Should I submit some Python code that might work as a buildmaster config to execute such a test?  Could someone give me the buildmaster config for a similar test from http://www.divmod.org/trac/ticket/2630 ?  It is the builder named "q-nevowinstall".

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_21"></a>|@zooko commented|
|-|-|

Sorry, I meant comment:16 is my suggestion for a test for this ticket.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun commented|
|-|-|

Hopefully those two files will help.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_23"></a>|@zooko commented|
|-|-|

[#3696](https://github.com/twisted/twisted/issues/3696) (Add extras_require to top-level setup.py to list Twisted optional dependancies) seems to be about implementing a superset of this ticket.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_24"></a>|@zooko set owner to @exarkun|
|-|-|

```twisted-master.cfg``` says:

```
from twisted_factories import (
    TwistedDocumentationBuildFactory, FullTwistedBuildFactory,
    GoodTwistedBuildFactory, TwistedReactorsBuildFactory, Win32RemovePYCs,
    LinuxPyOpenSSLBuildFactory, DebianPyOpenSSLBuildFactory,
    OSXPyOpenSSLBuildFactory, Win32PyOpenSSLBuildFactory,
    TwistedEasyInstallFactory, TwistedPyPyBuildFactory,
    PyPyTranslationFactory, TwistedIronPythonBuildFactory,
    TwistedScheduler)
```

So I think I need a ```twisted_factories.py``` file in order to understand how ```TwistedEasyInstallFactory``` works and propose a patch to it or a new factory modelled on it which implements the test described in comment:16.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_25"></a>|@zooko commented|
|-|-|

Opened [#4438](https://github.com/twisted/twisted/issues/4438) which emphasizes that if Twisted fails to declare an existing dependency then it is a bug, but if Twisted fails to declare a new dependency then it is a regression.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_26"></a>|@zooko commented|
|-|-|

Opened [#4438](https://github.com/twisted/twisted/issues/4438) which emphasizes that if Twisted fails to declare an existing dependency then it is a bug, but if Twisted fails to declare a new dependency then it is a regression.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_27"></a>|@zooko commented|
|-|-|

sorry about the accidental repetition--i'm juggling a baby with my other hand.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_29"></a>|@zooko commented|
|-|-|

Could someone who has access to the Twisted buildmaster config please share ```twisted_factories``` with me?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_30"></a>|@exarkun commented|
|-|-|

There you go.

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_31"></a>|@rodrigc set status to `closed`|
|-|-|

pypiwin32 was added as a dependency for Windows in [ticket:3696]
